integration/cluster.go

```diff
@@ -0,0 +1,224 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"time"
+
+	multierr "github.com/hashicorp/go-multierror"
+	errwrap "github.com/pkg/errors"
+)
+
+// Cluster represents an mgmt cluster. It uses the instance building blocks to
+// run clustered tests.
+type Cluster struct {
+	// Hostnames is the list of unique identifiers for this cluster.
+	Hostnames []string
+
+	// Preserve prevents the runtime output from being explicitly deleted.
+	// This is helpful for running analysis or tests on the output.
+	Preserve bool
+
+	// Debug enables more verbosity.
+	Debug bool
+
+	// dir is the directory where all files will be written under.
+	dir string
+
+	instances map[string]*Instance
+}
+
+// Init runs some initialization for this Cluster. It errors if the struct was
+// populated in an invalid way, or if it can't initialize correctly.
+func (obj *Cluster) Init() error {
+	obj.instances = make(map[string]*Instance)
+
+	// create temporary directory to use during testing
+	var err error
+	if obj.dir == "" {
+		obj.dir, err = ioutil.TempDir("", "mgmt-integration-cluster-")
+		if err != nil {
+			return errwrap.Wrapf(err, "can't create temporary directory")
+		}
+	}
+
+	for _, h := range obj.Hostnames {
+		instancePrefix := path.Join(obj.dir, h)
+		if err := os.MkdirAll(instancePrefix, dirMode); err != nil {
+			return errwrap.Wrapf(err, "can't create instance directory")
+		}
+
+		obj.instances[h] = &Instance{
+			Hostname: h,
+			Preserve: obj.Preserve,
+			Debug:    obj.Debug,
+
+			dir: instancePrefix,
+		}
+		if e := obj.instances[h].Init(); e != nil {
+			err = multierr.Append(err, e)
+		}
+	}
+
+	return err
+}
+
+// Close cleans up after we're done with this Cluster.
+func (obj *Cluster) Close() error {
+	var err error
+	// do this in reverse for fun
+	for i := len(obj.Hostnames) - 1; i >= 0; i-- {
+		h := obj.Hostnames[i]
+		instance, exists := obj.instances[h]
+		if !exists {
+			continue
+		}
+		if e := instance.Close(); e != nil {
+			err = multierr.Append(err, e)
+		}
+	}
+	if !obj.Preserve {
+		if obj.dir == "" || obj.dir == "/" {
```

Comment:
We might consider any path not created by TempDir to be dangerous, it's a user error of course, as the directory for this should only be a testrun specific tmpdir anyways. But still?

Reply:
Also why fail here and not during Init when the directory is set?

Reply:
Agreed. We could also add that, however this is just a "bonus" safety check because I'm a scaredy cat.